File: android/src/main/java/com/daboss2003/capacitor/imagemarker/PaddingResolver.kt
package com.daboss2003.capacitor.imagemarker

data class ResolvedInsets(
    val left: Float = 0f,
    val top: Float = 0f,
    val right: Float = 0f,
    val bottom: Float = 0f,
)

sealed class SpreadValue {
    data class Px(val value: Float) : SpreadValue()
    data class Percent(val value: Float) : SpreadValue()

    fun resolve(relativeTo: Float): Float = when (this) {
        is Px -> value
        is Percent -> value / 100f * relativeTo
    }

    companion object {
        fun parse(raw: Any?): SpreadValue? {
            return when (raw) {
                is Number -> Px(raw.toFloat())
                is String -> {
                    val trimmed = raw.trim()
                    if (trimmed.endsWith("%")) {
                        trimmed.dropLast(1).toFloatOrNull()?.let { Percent(it) }
                    } else {
                        trimmed.toFloatOrNull()?.let { Px(it) }
                    }
                }
                else -> null
            }
        }
    }
}

object PaddingResolver {
    fun resolve(source: Map<String, Any?>?, width: Float, height: Float): ResolvedInsets {
        if (source == null) return ResolvedInsets()

        var top: SpreadValue = SpreadValue.Px(0f)
        var right: SpreadValue = SpreadValue.Px(0f)
        var bottom: SpreadValue = SpreadValue.Px(0f)
        var left: SpreadValue = SpreadValue.Px(0f)

        source["padding"]?.let { raw ->
            val tokens = parseShorthand(raw)
            val expanded = expand(tokens)
            top = expanded[0]; right = expanded[1]; bottom = expanded[2]; left = expanded[3]
        }
        (source["paddingHorizontal"] ?: source["paddingX"])?.let { v ->
            SpreadValue.parse(v)?.let { left = it; right = it }
        }
        (source["paddingVertical"] ?: source["paddingY"])?.let { v ->
            SpreadValue.parse(v)?.let { top = it; bottom = it }
        }
        SpreadValue.parse(source["paddingLeft"])?.let { left = it }
        SpreadValue.parse(source["paddingRight"])?.let { right = it }
        SpreadValue.parse(source["paddingTop"])?.let { top = it }
        SpreadValue.parse(source["paddingBottom"])?.let { bottom = it }

        return ResolvedInsets(
            left = left.resolve(width),
            top = top.resolve(height),
            right = right.resolve(width),
            bottom = bottom.resolve(height),
        )
    }

    private fun parseShorthand(raw: Any): List<SpreadValue> {
        if (raw is Number) return listOf(SpreadValue.Px(raw.toFloat()))
        val str = raw as? String ?: return emptyList()
        return str.trim().split(Regex("\\s+")).mapNotNull { SpreadValue.parse(it) }
    }

    private fun expand(parts: List<SpreadValue>): List<SpreadValue> {
        val zero = SpreadValue.Px(0f)
        return when (parts.size) {
            0 -> listOf(zero, zero, zero, zero)
            1 -> listOf(parts[0], parts[0], parts[0], parts[0])
            2 -> listOf(parts[0], parts[1], parts[0], parts[1])
            3 -> listOf(parts[0], parts[1], parts[2], parts[1])
            else -> listOf(parts[0], parts[1], parts[2], parts[3])
        }
    }
}

File: ios/Plugin/PaddingResolver.swift
import Foundation
import CoreGraphics

struct ResolvedInsets {
    var left: CGFloat = 0
    var top: CGFloat = 0
    var right: CGFloat = 0
    var bottom: CGFloat = 0
}

enum SpreadValue {
    case px(CGFloat)
    case percent(CGFloat)

    func resolve(relativeTo dimension: CGFloat) -> CGFloat {
        switch self {
        case .px(let v):
            return v
        case .percent(let v):
            return v / 100.0 * dimension
        }
    }

    static func parse(_ raw: Any?) -> SpreadValue? {
        if let n = raw as? NSNumber {
            return .px(CGFloat(truncating: n))
        }
        guard let str = raw as? String else { return nil }
        let trimmed = str.trimmingCharacters(in: .whitespaces)
        if trimmed.hasSuffix("%") {
            let body = trimmed.dropLast()
            if let v = Double(body) { return .percent(CGFloat(v)) }
        } else if let v = Double(trimmed) {
            return .px(CGFloat(v))
        }
        return nil
    }
}

enum PaddingResolver {
    /// Resolves a CSS-like padding spec from a dictionary into pixel insets.
    /// Recognised keys: `padding` (shorthand string or number),
    /// `paddingLeft|paddingRight|paddingTop|paddingBottom`,
    /// `paddingHorizontal|paddingVertical`, `paddingX|paddingY`.
    static func resolve(from dict: [String: Any]?, width: CGFloat, height: CGFloat) -> ResolvedInsets {
        guard let dict = dict else { return ResolvedInsets() }

        var top: SpreadValue = .px(0)
        var right: SpreadValue = .px(0)
        var bottom: SpreadValue = .px(0)
        var left: SpreadValue = .px(0)

        if let raw = dict["padding"] {
            let tokens = shorthandTokens(raw)
            (top, right, bottom, left) = expandShorthand(tokens)
        }
        if let v = SpreadValue.parse(dict["paddingHorizontal"] ?? dict["paddingX"]) {
            left = v; right = v
        }
        if let v = SpreadValue.parse(dict["paddingVertical"] ?? dict["paddingY"]) {
            top = v; bottom = v
        }
        if let v = SpreadValue.parse(dict["paddingLeft"]) { left = v }
        if let v = SpreadValue.parse(dict["paddingRight"]) { right = v }
        if let v = SpreadValue.parse(dict["paddingTop"]) { top = v }
        if let v = SpreadValue.parse(dict["paddingBottom"]) { bottom = v }

        return ResolvedInsets(
            left: left.resolve(relativeTo: width),
            top: top.resolve(relativeTo: height),
            right: right.resolve(relativeTo: width),
            bottom: bottom.resolve(relativeTo: height)
        )
    }

    private static func shorthandTokens(_ raw: Any) -> [SpreadValue] {
        if let n = raw as? NSNumber {
            let v: SpreadValue = .px(CGFloat(truncating: n))
            return [v]
        }
        guard let str = raw as? String else { return [.px(0)] }
        return str.split(whereSeparator: { $0.isWhitespace }).compactMap { SpreadValue.parse(String($0)) }
    }

    private static func expandShorthand(_ parts: [SpreadValue]) -> (SpreadValue, SpreadValue, SpreadValue, SpreadValue) {
        switch parts.count {
        case 0:
            let z: SpreadValue = .px(0)
            return (z, z, z, z)
        case 1:
            return (parts[0], parts[0], parts[0], parts[0])
        case 2:
            return (parts[0], parts[1], parts[0], parts[1])
        case 3:
            return (parts[0], parts[1], parts[2], parts[1])
        default:
            return (parts[0], parts[1], parts[2], parts[3])
        }
    }
}
